src/net.cpp

```diff
@@ -408,9 +409,32 @@ CNode* CConnman::FindNode(const CService& addr)
     return nullptr;
 }
 
-bool CConnman::AlreadyConnectedToAddress(const CAddress& addr)
+bool CConnman::AlreadyConnectedToAddress(const CAddress& addr, ConnectionType conn_type, bool log)
 {
-    return FindNode(static_cast<CNetAddr>(addr)) || FindNode(addr.ToStringAddrPort());
+    if (Params().IsTestChain()) return false; // allow bilateral connections on test chains
+    const CService service{MaybeFlipIPv6toCJDNS(addr)};
+    const bool inbound{conn_type == ConnectionType::INBOUND};
+    CNode* pnode{nullptr};
+    if (inbound) {
+        // For accepting an inbound connection, check all existing connections by addr only.
+        pnode = FindNode(static_cast<CNetAddr>(service));
+    } else {
+        // For making an outbound connection, check the existing outbound connections
+        // by addr and port, and if none, then inbound connections by addr only.
+        pnode = FindNode(service.ToStringAddrPort(), /*inbound=*/false);
+        if (!pnode) pnode = FindNode(static_cast<CNetAddr>(service), /*inbound=*/true);
+    }
+    if (pnode && log) {
+        LogPrintLevel(BCLog::NET, inbound ? BCLog::Level::Debug : BCLog::Level::Info,
+                      "Not %s new %s connection%s, already connected to %s %s %s peer=%d%s, version=%d, subver=%s\n",
+                      inbound ? "accepting" : "opening", ConnectionTypeAsString(conn_type),
+                      fLogIPs ? strprintf(" %s %s", inbound ? "from" : "to", addr.ToStringAddrPort()) : "",
+                      TransportTypeAsString(pnode->m_transport->GetInfo().transport_type),
+                      pnode->ConnectionTypeAsString(), GetNetworkName(pnode->ConnectedThroughNetwork()),
```

Comment:
Dereferencing `pnode` after `FindNode()` has released `m_nodes_mutex` is not safe. I don't like adding recursive mutex calls, but without altering `FindNode()` we have to do that - this code has to acquire the mutex before any call to `FindNode()`, dereference the returned pointer and then release the mutex:

```diff
 bool CConnman::AlreadyConnectedToAddress(const CAddress& addr, ConnectionType conn_type, bool log)
 {
     if (Params().IsTestChain()) return false; // allow bilateral connections on test chains
     const CService service{MaybeFlipIPv6toCJDNS(addr)};
     const bool inbound{conn_type == ConnectionType::INBOUND};
     CNode* pnode{nullptr};
+    LOCK(m_nodes_mutex);
     if (inbound) {
         // For accepting an inbound connection, check all existing connections by addr only.
         pnode = FindNode(static_cast<CNetAddr>(service));
     } else {
         // For making an outbound connection, check the existing outbound connections
         // by addr and port, and if none, then inbound connections by addr only.
         pnode = FindNode(service.ToStringAddrPort(), /*inbound=*/false);
         if (!pnode) pnode = FindNode(static_cast<CNetAddr>(service), /*inbound=*/true);
     }
     if (pnode && log) {
         LogPrintLevel(BCLog::NET, inbound ? BCLog::Level::Debug : BCLog::Level::Info,
                       "Not %s new %s connection%s, already connected to %s %s %s peer=%d%s, version=%d, subver=%s\n",
                       inbound ? "accepting" : "opening", ConnectionTypeAsString(conn_type),
                       fLogIPs ? strprintf(" %s %s", inbound ? "from" : "to", addr.ToStringAddrPort()) : "",
                       TransportTypeAsString(pnode->m_transport->GetInfo().transport_type),
                       pnode->ConnectionTypeAsString(), GetNetworkName(pnode->ConnectedThroughNetwork()),
                       pnode->GetId(), fLogIPs ? strprintf(", peeraddr=%s", pnode->addr.ToStringAddrPort()) : "",
                       pnode->nVersion.load(), WITH_LOCK(pnode->m_subver_mutex, return pnode->cleanSubVer));
     }
-    return pnode;
+    return pnode != nullptr;
 }
```